[panicking] `Schema.Properties` gets inconsistent with the private `Schema.propertyNames`.

Hi there,

I have the following types

**main.go**

```go
type (
	UpsertRequest[D any] struct {
		Tag       string `json:"tag"`
		InputData D      `json:"input_data"`
	}

	User struct {
		Name    string `json:"name" minLength:"1" maxLength:"32"`
		Surname string `json:"surname" minLength:"1" maxLength:"32"`
	}

	noInputData = map[string]interface{}

	CreateUserInput struct {
		Body UpsertRequest[User]
	}

	CreateUserOutput struct {
		Body struct {
			Message string `json:"message"`
		}
	}

	updateUserRequest struct {
		UpsertRequest[noInputData]
	}

	UpdateUserInput struct {
		Body updateUserRequest
	}

	UpdateUserOutput struct {
		Body struct {
			Message string `json:"message"`
		}
	}
)

var _ huma.SchemaTransformer = updateUserRequest{}

func (i updateUserRequest) TransformSchema(r huma.Registry, s *huma.Schema) *huma.Schema {
	s.Description = "UpdateUserInput"
	delete(s.Properties, "input_data")
	return s
}

func addRoutes(api huma.API) {
	huma.Register(api, huma.Operation{
		OperationID: "CreateUser",
		Method:      http.MethodPost,
		Path:        "/user",
	}, func(ctx context.Context, input *CreateUserInput) (*CreateUserOutput, error) {
		resp := &CreateUserOutput{}
		resp.Body.Message = "CreateUser works!"
		return resp, nil
	})

	huma.Register(api, huma.Operation{
		OperationID: "UpdateUser",
		Method:      http.MethodPut,
		Path:        "/user",
	}, func(ctx context.Context, input *UpdateUserInput) (*UpdateUserOutput, error) {
		resp := &UpdateUserOutput{}
		resp.Body.Message = "UpdateUser works!"
		return resp, nil
	})
}
```

Here's the full generated spec.

<details>
<summary>spec</summary>

```yaml
components:
  schemas:
    CreateUserOutputBody:
      additionalProperties: false
      properties:
        $schema:
          description: A URL to the JSON Schema for this object.
          examples:
            - https://example.com/schemas/CreateUserOutputBody.json
          format: uri
          readOnly: true
          type: string
        message:
          type: string
      required:
        - message
      type: object
    ErrorDetail:
      additionalProperties: false
      properties:
        location:
          description: Where the error occurred, e.g. 'body.items[3].tags' or 'path.thing-id'
          type: string
        message:
          description: Error message text
          type: string
        value:
          description: The value at the given location
      type: object
    ErrorModel:
      additionalProperties: false
      properties:
        $schema:
          description: A URL to the JSON Schema for this object.
          examples:
            - https://example.com/schemas/ErrorModel.json
          format: uri
          readOnly: true
          type: string
        detail:
          description: A human-readable explanation specific to this occurrence of the problem.
          examples:
            - Property foo is required but is missing.
          type: string
        errors:
          description: Optional list of individual error details
          items:
            $ref: "#/components/schemas/ErrorDetail"
          type:
            - array
            - "null"
        instance:
          description: A URI reference that identifies the specific occurrence of the problem.
          examples:
            - https://example.com/error-log/abc123
          format: uri
          type: string
        status:
          description: HTTP status code
          examples:
            - 400
          format: int64
          type: integer
        title:
          description: A short, human-readable summary of the problem type. This value should not change between occurrences of the error.
          examples:
            - Bad Request
          type: string
        type:
          default: about:blank
          description: A URI reference to human-readable documentation for the error.
          examples:
            - https://example.com/errors/example
          format: uri
          type: string
      type: object
    UpdateUserOutputBody:
      additionalProperties: false
      properties:
        $schema:
          description: A URL to the JSON Schema for this object.
          examples:
            - https://example.com/schemas/UpdateUserOutputBody.json
          format: uri
          readOnly: true
          type: string
        message:
          type: string
      required:
        - message
      type: object
    UpdateUserRequest:
      additionalProperties: false
      description: UpdateUserInput
      properties:
        $schema:
          description: A URL to the JSON Schema for this object.
          examples:
            - https://example.com/schemas/UpdateUserRequest.json
          format: uri
          readOnly: true
          type: string
        tag:
          type: string
      required:
        - tag
        - input_data
      type: object
    UpsertRequestUser:
      additionalProperties: false
      properties:
        $schema:
          description: A URL to the JSON Schema for this object.
          examples:
            - https://example.com/schemas/UpsertRequestUser.json
          format: uri
          readOnly: true
          type: string
        input_data:
          $ref: "#/components/schemas/User"
        tag:
          type: string
      required:
        - tag
        - input_data
      type: object
    User:
      additionalProperties: false
      properties:
        name:
          maxLength: 32
          minLength: 1
          type: string
        surname:
          maxLength: 32
          minLength: 1
          type: string
      required:
        - name
        - surname
      type: object
info:
  title: My API
  version: 1.0.0
openapi: 3.1.0
paths:
  /user:
    post:
      operationId: CreateUser
      requestBody:
        content:
          application/json:
            schema:
              $ref: "#/components/schemas/UpsertRequestUser"
        required: true
      responses:
        "200":
          content:
            application/json:
              schema:
                $ref: "#/components/schemas/CreateUserOutputBody"
          description: OK
        default:
          content:
            application/problem+json:
              schema:
                $ref: "#/components/schemas/ErrorModel"
          description: Error
    put:
      operationId: UpdateUser
      requestBody:
        content:
          application/json:
            schema:
              $ref: "#/components/schemas/UpdateUserRequest"
        required: true
      responses:
        "200":
          content:
            application/json:
              schema:
                $ref: "#/components/schemas/UpdateUserOutputBody"
          description: OK
        default:
          content:
            application/problem+json:
              schema:
                $ref: "#/components/schemas/ErrorModel"
          description: Error


```
</details>

In this case, I want to have two operations `Create` and `Update` that:
- Share the data structures ( `UpsertRequest[D any]` particularly ) 
- But `Update` doesn't have one of the fields ( `input_data` ) on the input.

To avoid duplication of the types ( IRL, these types are quite huge and complex ), I reuse them under `CreateUserInput` and `updateUserRequest` ( a wrapper for a human-readable schema name ) + `UpdateUserInput`.

`updateUserRequest` has `TransformSchema` that removes `input_data` from `s.Properties`

Everything works perfectly for the OAS generation.

**The issue**

main_test.go

```go
func TestUserCreate(t *testing.T) {
	_, api := humatest.New(t, huma.DefaultConfig("Test API", "1.0.0"))
	addRoutes(api)

	// Create a new user
	_ = api.Post("/user", map[string]any{"tag": "tag1", "input_data": map[string]any{"name": "John", "surname": "Doe"}})

	// Update the user
	_ = api.Put("/user", map[string]any{"tag": "tag1", "input_data": map[string]any{"name": "John", "surname": "Doe"}})
}
```

I get a panic on the `PUT /user` request with `input_data` ( that I expect shouldn't pass the validation )

```sh
$> go test ./...

--- FAIL: TestUserCreate (0.00s)
    main_test.go:15: Making request:
        POST /user HTTP/1.1
        Content-Type: application/json
        
        {
          "input_data": {
            "name": "John",
            "surname": "Doe"
          },
          "tag": "tag1"
        }
    main_test.go:15: Got response:
        HTTP/1.1 200 OK
        Connection: close
        Content-Type: application/json
        Link: </schemas/UpdateUserOutputBody.json>; rel="describedBy"
        
        {
          "$schema": "https:///schemas/UpdateUserOutputBody.json",
          "message": "CreateUser works!"
        }
    main_test.go:18: Making request:
        PUT /user HTTP/1.1
        Content-Type: application/json
        
        {
          "input_data": {
            "name": "John",
            "surname": "Doe"
          },
          "tag": "tag1"
        }
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x161 pc=0x5f20ca]

goroutine 6 [running]:
testing.tRunner.func1.2({0x64ae60, 0x882c00})
	/home/superstas/.gvm/pkgsets/go1.21.13/global/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.8.linux-amd64/src/testing/testing.go:1631 +0x24a
testing.tRunner.func1()
	/home/superstas/.gvm/pkgsets/go1.21.13/global/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.8.linux-amd64/src/testing/testing.go:1634 +0x377
panic({0x64ae60?, 0x882c00?})
	/home/superstas/.gvm/pkgsets/go1.21.13/global/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.8.linux-amd64/src/runtime/panic.go:770 +0x132
github.com/danielgtaylor/huma/v2.handleMapString({0x7001c0, 0xc000016e00}, 0xc00018ef08, 0xc000078c60, 0x1, 0xc0001d4cc0, 0xc000012558)
	/home/superstas/.gvm/pkgsets/go1.21.13/global/pkg/mod/github.com/danielgtaylor/huma/v2@v2.23.0/validate.go:595 +0x4ca
github.com/danielgtaylor/huma/v2.Validate({0x7001c0, 0xc000016e00}, 0x6fe080?, 0xc000078c60, 0x1, {0x6489a0, 0xc0001d4cc0}, 0xc000012558)
	/home/superstas/.gvm/pkgsets/go1.21.13/global/pkg/mod/github.com/danielgtaylor/huma/v2@v2.23.0/validate.go:524 +0xbf4
github.com/danielgtaylor/huma/v2.Register[...].func1()
	/home/superstas/.gvm/pkgsets/go1.21.13/global/pkg/mod/github.com/danielgtaylor/huma/v2@v2.23.0/huma.go:1268 +0x1b45
```


My internal research showed that modifying `s.Properties` directly makes it inconsistent with the private `propertyNames` ( [ref](https://github.com/danielgtaylor/huma/blob/main/schema.go#L132C2-L132C15) ).

I fixed this internally by this patch
```diff
diff --git a/validate.go b/validate.go
index 365a237..193f295 100644
--- a/validate.go
+++ b/validate.go
@@ -586,7 +586,11 @@ func handleMapString(r Registry, s *Schema, path *PathBuffer, mode ValidateMode,
 	}
 
 	for _, k := range s.propertyNames {
-		v := s.Properties[k]
+		v, ok := s.Properties[k]
+		if !ok {
+			continue
+		}
 
 		// Schemas are generated such that the read/write-only properties are set
 		// alongside the `$ref`, if it is present (i.e. for objects). If not,
@@ -680,7 +684,11 @@ func handleMapAny(r Registry, s *Schema, path *PathBuffer, mode ValidateMode, m
 	}
 
 	for _, k := range s.propertyNames {
-		v := s.Properties[k]
+		v, ok := s.Properties[k]
+		if !ok {
+			continue
+		}
 
 		// Schemas are generated such that the read/write-only properties are set
 		// alongside the `$ref`, if it is present (i.e. for objects). If not,
```

But there are a few questions:
- Is it a bug or the wrong way of using Huma?
- What is the best way to achieve the state I was trying to achieve? ( having a shared Go structure but slightly different OAS + Validation )


Thank you.

--- a/huma_test.go
+++ b/huma_test.go
@@ -1951,13 +1951,13 @@ func TestExhaustiveErrors(t *testing.T) {
 				"location": "path.id",
 				"value": 15
 			}, {
-				"message": "expected length <= 10",
-				"location": "body.name",
-				"value": "12345678901"
-			}, {
 				"message": "expected number >= 1",
 				"location": "body.count",
 				"value": -6
+			}, {
+				"message": "expected length <= 10",
+				"location": "body.name",
+				"value": "12345678901"
 			}, {
 				"message": "input resolver error",
 				"location": "path.id",

--- a/schema.go
+++ b/schema.go
@@ -290,18 +290,15 @@ func (s *Schema) PrecomputeMessages() {
 		}
 	}
 
-	if s.propertyNames == nil {
-		s.propertyNames = make([]string, 0, len(s.Properties))
-		for name := range s.Properties {
-			s.propertyNames = append(s.propertyNames, name)
-		}
+	s.propertyNames = make([]string, 0, len(s.Properties))
+	for name := range s.Properties {
+		s.propertyNames = append(s.propertyNames, name)
 	}
+	sort.Strings(s.propertyNames)
 
-	if s.requiredMap == nil {
-		s.requiredMap = map[string]bool{}
-		for _, name := range s.Required {
-			s.requiredMap[name] = true
-		}
+	s.requiredMap = map[string]bool{}
+	for _, name := range s.Required {
+		s.requiredMap[name] = true
 	}
 
 	if s.Items != nil {
@@ -672,7 +669,10 @@ func SchemaFromType(r Registry, t reflect.Type) *Schema {
 	// Transform generated schema if type implements SchemaTransformer
 	v := reflect.New(t).Interface()
 	if st, ok := v.(SchemaTransformer); ok {
-		return st.TransformSchema(r, s)
+		s = st.TransformSchema(r, s)
+
+		// The schema may have been modified, so recompute the error messages.
+		s.PrecomputeMessages()
 	}
 	return s
 }

--- a/validate_test.go
+++ b/validate_test.go
@@ -1448,7 +1448,7 @@ func TestValidateSchemaTransformerDeleteField(t *testing.T) {
 
 	huma.Validate(registry, s, pb, huma.ModeReadFromServer, map[string]any{"field1": "value1"}, res)
 	// We should have no errors and no panics.
-	assert.Len(t, res.Errors, 0)
+	assert.Empty(t, res.Errors)
 }
 
 func ExampleModelValidator() {

--- a/validate_test.go
+++ b/validate_test.go
@@ -1449,6 +1449,7 @@ func TestValidateSchemaTransformerDeleteField(t *testing.T) {
 	huma.Validate(registry, s, pb, huma.ModeReadFromServer, map[string]any{"field1": "value1"}, res)
 	// We should have no errors and no panics.
 	assert.Empty(t, res.Errors)
+	assert.NotContains(t, s.Properties, "field2")
 }
 
 func ExampleModelValidator() {

--- a/validate_test.go
+++ b/validate_test.go
@@ -1428,6 +1428,29 @@ func TestValidateCustomFormatter(t *testing.T) {
 	assert.Equal(t, "custom: [mail: missing '@' or angle-addr] (value: alice)", res.Errors[0].Error())
 }
 
+type TransformDeleteField struct {
+	Field1 string `json:"field1"`
+	Field2 string `json:"field2"`
+}
+
+var _ huma.SchemaTransformer = (*TransformDeleteField)(nil)
+
+func (t *TransformDeleteField) TransformSchema(r huma.Registry, s *huma.Schema) *huma.Schema {
+	delete(s.Properties, "field2")
+	return s
+}
+
+func TestValidateSchemaTransformerDeleteField(t *testing.T) {
+	registry := huma.NewMapRegistry("#/components/schemas/", huma.DefaultSchemaNamer)
+	s := registry.Schema(reflect.TypeOf(&TransformDeleteField{}), true, "TestInput")
+	pb := huma.NewPathBuffer([]byte(""), 0)
+	res := &huma.ValidateResult{}
+
+	huma.Validate(registry, s, pb, huma.ModeReadFromServer, map[string]any{"field1": "value1"}, res)
+	// We should have no errors and no panics.
+	assert.Len(t, res.Errors, 0)
+}
+
 func ExampleModelValidator() {
 	// Define a type you want to validate.
 	type Model struct {
@@ -1452,7 +1475,7 @@ func ExampleModelValidator() {
 	errs = validator.Validate(typ, val)
 	fmt.Println(errs)
 
-	// Output: [expected length <= 5 (name: abcdefg) expected number >= 25 (age: 1)]
+	// Output: [expected number >= 25 (age: 1) expected length <= 5 (name: abcdefg)]
 	// []
 }
 

